### ISSUE TITLE
Rust doesn't build on Ubuntu 12.04

### ISSUE THREAD

**Issue by brson** [rust-lang/rust#14023]
It doesn't have g++ > 4.7, which LLVM needs. This is apparently breaking http://rust-ci.org/.

**Comment by emanueLczirai** [comment]
Hi. I'm having this issue as well, on Linux Manjaro (not ubuntu) with (newer than 4.7) gcc == 4.9.2
There are two results when searching for `cstddef`:

```
/usr/include/boost/compatibility/cpp_c_headers/cstddef
/usr/include/c++/4.9.2/cstddef
```

This is my gcc version 4.9.2:
```
gcc -v
Using built-in specs.
COLLECT_GCC=gcc
COLLECT_LTO_WRAPPER=/usr/lib/gcc/x86_64-unknown-linux-gnu/4.9.2/lto-wrapper
Target: x86_64-unknown-linux-gnu
Configured with: /build/gcc/src/gcc-4.9-20141224/configure --prefix=/usr --libdir=/usr/lib --libexecdir=/usr/lib --mandir=/usr/share/man --infodir=/usr/share/info --with-bugurl=https://bugs.archlinux.org/ --enable-languages=c,c++,ada,fortran,go,lto,objc,obj-c++ --enable-shared --enable-threads=posix --with-system-zlib --enable-__cxa_atexit --disable-libunwind-exceptions --enable-clocale=gnu --disable-libstdcxx-pch --disable-libssp --enable-gnu-unique-object --enable-linker-build-id --enable-cloog-backend=isl --enable-lto --enable-plugin --enable-install-libiberty --with-linker-hash-style=gnu --disable-multilib --disable-werror --enable-checking=release
Thread model: posix
gcc version 4.9.2 20141224 (prerelease) (GCC) 
```

Trying to compile latest rust from master with something like this:

```
$ ./configure --prefix=/usr --disable-rpath --disable-verify-install --disable-debug --enable-manage-submodules --enable-clang --enable-ccache --enable-dist-host-only 
$ make -j1 all
```

```
...
make[2]: Entering directory '/home/emacs/build/rust-git/makepkg/rust-git/src/rust/x86_64-unknown-linux-gnu/llvm/lib/Support'
/home/emacs/build/rust-git/makepkg/rust-git/src/rust/src/llvm/autoconf/mkinstalldirs /home/emacs/build/rust-git/makepkg/rust-git/src/rust/x86_64-unknown-linux-gnu/llvm/lib/Support/Release+Asserts > /dev/null
/usr/bin/date > /home/emacs/build/rust-git/makepkg/rust-git/src/rust/x86_64-unknown-linux-gnu/llvm/lib/Support/Release+Asserts/.dir
llvm[2]: Compiling APFloat.cpp for Release+Asserts build
if  ccache clang++ -Qunused-arguments -I/home/emacs/build/rust-git/makepkg/rust-git/src/rust/src/llvm/include -I/home/emacs/build/rust-git/makepkg/rust-git/src/rust/src/llvm/lib/Support -I/home/emacs/build/rust-git/makepkg/rust-git/src/rust/x86_64-unknown-linux-gnu/llvm/include -I/home/emacs/build/rust-git/makepkg/rust-git/src/rust/x86_64-unknown-linux-gnu/llvm/lib/Support  -D_DEBUG -D_GNU_SOURCE -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS -O3 -fomit-frame-pointer -stdlib=libc++ -std=c++11 -fvisibility-inlines-hidden -fno-exceptions -fno-rtti -fPIC -ffunction-sections -fdata-sections -Wcast-qual -march=native -O2 -pipe -fstack-protector-strong --param=ssp-buffer-size=4 -D_FORTIFY_SOURCE=2 -Wno-trigraphs   -D_FORTIFY_SOURCE=2 -D_FORTIFY_SOURCE=2 -D_FORTIFY_SOURCE=2  -Wall -W -Wno-unused-parameter -Wwrite-strings -Wcovered-switch-default -Wno-uninitialized -Wno-missing-field-initializers -c -MMD -MP -MF "/home/emacs/build/rust-git/makepkg/rust-git/src/rust/x86_64-unknown-linux-gnu/llvm/lib/Support/Release+Asserts/APFloat.d.tmp" -MT "/home/emacs/build/rust-git/makepkg/rust-git/src/rust/x86_64-unknown-linux-gnu/llvm/lib/Support/Release+Asserts/APFloat.o" -MT "/home/emacs/build/rust-git/makepkg/rust-git/src/rust/x86_64-unknown-linux-gnu/llvm/lib/Support/Release+Asserts/APFloat.d" /home/emacs/build/rust-git/makepkg/rust-git/src/rust/src/llvm/lib/Support/APFloat.cpp -o /home/emacs/build/rust-git/makepkg/rust-git/src/rust/x86_64-unknown-linux-gnu/llvm/lib/Support/Release+Asserts/APFloat.o ; \
        then /usr/bin/mv -f "/home/emacs/build/rust-git/makepkg/rust-git/src/rust/x86_64-unknown-linux-gnu/llvm/lib/Support/Release+Asserts/APFloat.d.tmp" "/home/emacs/build/rust-git/makepkg/rust-git/src/rust/x86_64-unknown-linux-gnu/llvm/lib/Support/Release+Asserts/APFloat.d"; else /usr/bin/rm "/home/emacs/build/rust-git/makepkg/rust-git/src/rust/x86_64-unknown-linux-gnu/llvm/lib/Support/Release+Asserts/APFloat.d.tmp"; exit 1; fi
In file included from /home/emacs/build/rust-git/makepkg/rust-git/src/rust/src/llvm/lib/Support/APFloat.cpp:15:
In file included from /home/emacs/build/rust-git/makepkg/rust-git/src/rust/src/llvm/include/llvm/ADT/APFloat.h:20:
In file included from /home/emacs/build/rust-git/makepkg/rust-git/src/rust/src/llvm/include/llvm/ADT/APInt.h:19:
In file included from /home/emacs/build/rust-git/makepkg/rust-git/src/rust/src/llvm/include/llvm/ADT/ArrayRef.h:14:
/home/emacs/build/rust-git/makepkg/rust-git/src/rust/src/llvm/include/llvm/ADT/STLExtras.h:21:10: fatal error: 
      'cstddef' file not found
#include <cstddef> // for std::size_t
         ^
1 error generated.
/usr/bin/rm: cannot remove ‘/home/emacs/build/rust-git/makepkg/rust-git/src/rust/x86_64-unknown-linux-gnu/llvm/lib/Support/Release+Asserts/APFloat.d.tmp’: No such file or directory
/home/emacs/build/rust-git/makepkg/rust-git/src/rust/src/llvm/Makefile.rules:1514: recipe for target '/home/emacs/build/rust-git/makepkg/rust-git/src/rust/x86_64-unknown-linux-gnu/llvm/lib/Support/Release+Asserts/APFloat.o' failed
make[2]: *** [/home/emacs/build/rust-git/makepkg/rust-git/src/rust/x86_64-unknown-linux-gnu/llvm/lib/Support/Release+Asserts/APFloat.o] Error 1
make[2]: Leaving directory '/home/emacs/build/rust-git/makepkg/rust-git/src/rust/x86_64-unknown-linux-gnu/llvm/lib/Support'
/home/emacs/build/rust-git/makepkg/rust-git/src/rust/src/llvm/Makefile.rules:873: recipe for target 'all' failed
make[1]: *** [all] Error 1
make[1]: Leaving directory '/home/emacs/build/rust-git/makepkg/rust-git/src/rust/x86_64-unknown-linux-gnu/llvm'
/tmp/makepkg/rust-git/src/rust/mk/llvm.mk:60: recipe for target '/tmp/makepkg/rust-git/src/rust/x86_64-unknown-linux-gnu/llvm/Release+Asserts/bin/llvm-config' failed
make: *** [/tmp/makepkg/rust-git/src/rust/x86_64-unknown-linux-gnu/llvm/Release+Asserts/bin/llvm-config] Error 2
==> ERROR: A failure occurred in build().
    Aborting...

real	1m7.847s
user	0m42.023s
sys	0m10.023s

```

I'll let you know if I fix it.

**Comment by emanueLczirai** [comment]
tl;dr: fixed by passing `--disable-libcpp` to rust's ./configure

--------longer read:
found the culprit, turns out that by having `-stdlib=libc++` parameter passed to `clang++` it will fail as mentioned above with the `cstddef` error and simply removing it  works. But to properly remove it I have to either not specify `--enable-libcpp` (by default it's NO) or specify `--disable-libcpp`  to the `./configure` script of `llvm` (not of rust). This translates to having to pass `--disable-libcpp` to the rust `./configure`.

(tested to work as specified)

I also found this text, afterwards, to be related (from file: [GettingStarted.rst](https://github.com/llvm-mirror/llvm/blob/master/docs/GettingStarted.rst#getting-a-modern-host-c-toolchain)):
```rest
When you build Clang, you will need to give *it* access to modern C++11
standard library in order to use it as your new host in part of a bootstrap.
There are two easy ways to do this, either build (and install) libc++ along
with Clang and then use it with the ``-stdlib=libc++`` compile and link flag,
or install Clang into the same prefix (``$HOME/toolchains`` above) as GCC.
Clang will look within its own prefix for libstdc++ and use it if found. You
can also add an explicit prefix for Clang to look in for a GCC toolchain with
the ``--gcc-toolchain=/opt/my/gcc/prefix`` flag, passing it to both compile and
link commands when using your just-built-Clang to bootstrap.

```

**Comment by steveklabnik** [comment]
It seems like this is taken care of now, or at least, as much as it's going to be.
